### PR TITLE
Variable length message fields allocation free [7550] 

### PR DESCRIPTION
**Pull request by IkerLuengo**
Reworked to keep API compatibility

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1373/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1144/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1371/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1375/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1146/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1373/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1376/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1147/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1374/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1377/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1150/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1375/)

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 1]
```diff
@@ -58,9 +58,9 @@ ParticipantProxyData::ParticipantProxyData(
     , should_check_lease_duration(false)
     , m_readers(allocation.readers)
     , m_writers(allocation.writers)
-{
-    m_userData.max_size(static_cast<uint32_t>(allocation.data_limits.max_user_data));
-}
+    {
+        m_userData.set_max_size(static_cast<uint32_t>(allocation.data_limits.max_user_data));
+    }
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: Wrong indentation here.

**Comment by IkerLuengo** [comment]
> On both QosPolicies.h (public and mock), on the simple `clear` methods, I think it is better to have a constexpr and a copy than a stack local and a swap.

We cannot have a constexpr because the destructor of the classes is not trivial (as the base classes have virtual destructors)

**Comment by IkerLuengo** [comment]
We were sending UserData and Partitions always (even if empty) because 'hasChanged' was set to true on clear, instead of resetting it to its default value.

TypeConsistencyEnforcement reading would fail if size > 7.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1382/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1155/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1380/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1383/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1156/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1381/)

**Comment by dirk-thomas** [comment]
This change broke cross vendor compatibility with Connext: see #1014.

**Pull request merged**